### PR TITLE
Added teacher application admin management pages

### PR DESCRIPTION
**Pull request by aoby**
Depends on https://github.com/code-dot-org/code-dot-org/pull/14780

 1. Encapsulated teacher application email logic in a new class, `Pd::TeacherApplicationEmailParams`, from the scripts in `bin/mail/2017-teacher-application-responses` and updated tests accordingly.

 2. Added ability to edit and validate program registration (in Pegasus forms) from the `Pd::TeacherApplication` model.
 
 3. Added admin-only controller actions for listing and editing teacher applications, and sending decision emails.

### Manage
image

### Edit
image

#### followed by (for teacherCon)
image

#### Validations
image

#### teachercon
image

#### partner
image

#### waitlist or decline
image

#### Validations
image

### CODE REVIEW

File: dashboard/app/controllers/pd/teacher_application_controller.rb [hunk 1]
```diff
@@ -13,4 +22,96 @@ def new
...
+      # It's a valid id. Navigate directly to that application.
+      if query =~ /^\d+$/ && Pd::TeacherApplication.exists?(query.to_i)
+        redirect_to action: :edit, teacher_application_id: query
+        return
+      end
+
+      # Search emails
+      @teacher_applications = @teacher_applications.where(
+        'primary_email LIKE ? OR secondary_email LIKE ?',
+        "%#{query}%", "%#{query}%"
```
Comment (Hamms) [inline dashboard/app/controllers/pd/teacher_application_controller.rb:40]: even though this is an admin-only route, this still feels like we're just begging for an SQL injection. Is there a safer way we can do this?

File: dashboard/app/models/pd/teacher_application.rb [hunk 2]
```diff
@@ -260,4 +284,125 @@ def to_expanded_json
...
+  before_validation :save_move_to_user
+  def save_move_to_user
+    found_user = lookup_move_to_user
+    self.user = found_user if found_user
+  end
+
+  def lookup_move_to_user
+    return nil if move_to_user.blank?
+
+    if move_to_user.to_i.to_s == move_to_user
```
Comment (Hamms) [inline dashboard/app/models/pd/teacher_application.rb]: this seems like an unnecessarily roundabout way of confirming that the argument is an integer string

File: apps/style/code-studio/pd.scss [hunk 3]
```diff
@@ -34,6 +34,14 @@
     white-space: normal;
     font-weight: normal;
   }
+
```
Comment (mehalshah) [inline apps/style/code-studio/pd.scss:37]: Do you need all the rest of the stuff in pd.scss here? 

In any case, I know we are hoping to more tightly scope our CSS. If you don't want this in a separate file, scope them to your specific elements.

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb [hunk 4]
```diff
@@ -151,80 +151,57 @@ def load_workshop_map
 
   def process(decision, teacher_application, params = {})
     raise "Unexpected decision: #{decision}" unless @results.key? decision
+    params[:decision] = decision
```
Comment (mehalshah) [inline bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: Altering params passed in is bad practice. You could always do something like
`new_params = params.merge({decision: decision})`

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb [hunk 5]
```diff
@@ -151,80 +151,57 @@ def load_workshop_map
...
 
-    # Construct result hash, add to appropriate list, and return the new result
-    {
-      name: teacher_application.teacher_name,
-      email: teacher_application.primary_email,
-      preferred_first_name_s: teacher_application.teacher_first_name,
-      course_name_s: teacher_application.program_name
-    }.merge(params).tap do |result|
-      @results[decision] << result
+    email_params.to_final_params.except(:decision).tap do |effective_params|
```
Comment (mehalshah) [inline bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: Then here just reference params

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb [hunk 6]
```diff
@@ -151,80 +151,57 @@ def load_workshop_map
...
-    # Next, update the actual dashboard DB with this accepted workshop string.
-    save_accepted_workshop teacher_application, program, accepted_workshop
+    workshop_info = lookup_workshop teacher_application.accepted_workshop
 
     params = {
       regional_partner_name_s: workshop_info[:partner_name],
       regional_partner_contact_person_s: workshop_info[:partner_contact],
       regional_partner_contact_person_email_s: workshop_info[:partner_email],
-      workshop_registration_url_s: "https://studio.code.org/pd/workshops/#{workshop_info[:id]}/enroll",
+      workshop_registration_url_s: workshop_info[:id],
```
Comment (Hamms) [inline bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: I'm not sure I understand why this change is correct.

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb [hunk 7]
```diff
@@ -151,80 +151,57 @@ def load_workshop_map
...
-
   def process_accept(teacher_application, program, accepted_workshop, regional_partner_override)
+    # First, update the actual dashboard DB with this accepted workshop string.
+    save_accepted_workshop teacher_application, program, accepted_workshop, regional_partner_override
+
     # There are 2 kinds of acceptance, TeacherCon (ours) and Regional Partner.
-    # We can tell based on the accepted workshop
-    if teachercon? accepted_workshop
-      process_accept_teachercon teacher_application, program, accepted_workshop, regional_partner_override
+    if teacher_application.accepted_program.teachercon?
```
Comment (mehalshah) [inline bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb:169]: You can probably write this as a ternary. Your call

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb (hunk 6 above)
Comment (aoby) [reply bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: Because now the [transform](https://github.com/code-dot-org/code-dot-org/pull/14781/files#diff-80cf108bdbd3c072c73a5cd175b0b5fcR88) is applied in [TeacherApplicationEmailParams.to_final_params](https://github.com/code-dot-org/code-dot-org/pull/14781/files#diff-4fecf845c0f113ce052092a276fa45d0R93). Until then, the value is kept as a raw value so it can be displayed with a label in the UI. In this case, even though the final param key is `workshop_registration_url_s`, the label, and therefore the live value is [Workshop Id](https://github.com/code-dot-org/code-dot-org/pull/14781/files#diff-80cf108bdbd3c072c73a5cd175b0b5fcR85).

File: dashboard/app/controllers/pd/teacher_application_controller.rb [hunk 8]
```diff
@@ -13,4 +22,96 @@ def new
...
+  end
+
+  # PATCH /pd/teacher_application/manage/:teacher_application_id
+  # Admin only
+  def update
+    if @teacher_application.update application_params.merge(admin_managed: true)
+      flash[:notice] = "Teacher Application #{@teacher_application.id} saved."
+      redirect_to action: :edit
+    else
+      render :edit
```
Comment (mehalshah) [inline dashboard/app/controllers/pd/teacher_application_controller.rb:65]: Consider rendering the errors from the failed update

File: dashboard/app/controllers/pd/teacher_application_controller.rb (hunk 1 above)
Comment (aoby) [reply dashboard/app/controllers/pd/teacher_application_controller.rb:40]: This format with the `?` operator in the where clause, and the params passed as separate arguments, is meant to combat SQL injection. See `7.2.4 Countermeasures` here: http://guides.rubyonrails.org/security.html#sql-injection.

Do you think this is insufficient?

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb (hunk 6 above)
Comment (Hamms) [reply bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: In that case, can the name of the param be updated to something more like `workshop_id_i`?

File: dashboard/app/models/pd/accepted_program.rb [hunk 9]
```diff
@@ -20,4 +20,10 @@ class Pd::AcceptedProgram < ActiveRecord::Base
   validates_inclusion_of :course, in: Pd::TeacherApplication::PROGRAM_DETAILS_BY_COURSE.keys, if: -> {course.present?}
   validates_presence_of :user
   validates_uniqueness_of :user_id, scope: :course, message: 'already has an entry for this course'
+
+  TEACHER_CONS_CITY_REGEX = /Houston|Phoenix|Philadelphia/
+
+  def teachercon?
+    !!(workshop_name =~ TEACHER_CONS_CITY_REGEX)
```
Comment (mehalshah) [inline dashboard/app/models/pd/accepted_program.rb]: There's no guarantee that other workshops won't have the name Houston, Phoenix, or Philadelphia in them. Since I'm a huge geo nerd, I'll just point out that there is a Philadelphia Mississippi.

not sure right now what's the better way. Do we want to have a teachercon bit on the workshop model?

File: dashboard/app/models/pd/teacher_application.rb [hunk 10]
```diff
@@ -88,15 +94,27 @@ def validate_required_application_fields
 
   validate :primary_email_must_match_user_email
   def primary_email_must_match_user_email
-    if user && user.hashed_email != Digest::MD5.hexdigest(primary_email)
-      account_settings_link = ActionController::Base.helpers.link_to('account settings', '/users/edit')
-      errors.add(
-        :primary_email,
-        "must match your login. If you want to use this email instead, first update it in #{account_settings_link}."
+    return unless user
```
Comment (mehalshah) [inline dashboard/app/models/pd/teacher_application.rb:97]: This return is redundant - primary_email_matches_user_email? will skip this block

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb (hunk 6 above)
Comment (aoby) [reply bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb]: Per offline discussion, I'm going to change `transform` to output a new key and value. This will be in: `workshop_id_i`, out: `workshop_registration_url_s`

File: dashboard/app/controllers/pd/teacher_application_controller.rb (hunk 8 above)
Comment (aoby) [reply dashboard/app/controllers/pd/teacher_application_controller.rb:65]: Already [done](https://github.com/code-dot-org/code-dot-org/pull/14781/files#diff-34e089fcf4c547c0532a0ae4eff088d9R21) :)

File: dashboard/app/models/pd/accepted_program.rb (hunk 9 above)
Comment (aoby) [reply dashboard/app/models/pd/accepted_program.rb]: Good call. We're doing it two different ways now, using specific strings in the [email scripts](https://github.com/code-dot-org/code-dot-org/blob/staging/bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb#L48-L52) and this regex in the [program registration form](https://github.com/code-dot-org/code-dot-org/blob/staging/pegasus/sites.v3/code.org/public/pd-program-registration/index.haml#L8). I'll use the specific strings and put it in one place in a shared spot that both code paths can leverage. 

No, we don't have a teachercon bit on the workshop model, but it doesn't belong there anyway because teachercons don't have an associated workshop entry. Registration happens on the program registration form.

File: dashboard/app/models/pd/teacher_application.rb (hunk 10 above)
Comment (aoby) [reply dashboard/app/models/pd/teacher_application.rb:97]: It's not. `primary_email_matches_user_email?` will return true only if there is a user and its email matches, meaning it will return false if there is no user. Here I specifically don't want to fail the validation if there is no user.

File: dashboard/lib/pd/program_registration_validation.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+require File.join(CDO.root_dir, 'lib/forms/pegasus_form_validation')
+require File.join(CDO.root_dir, 'pegasus/forms/pd_program_registration')
```
Comment (mehalshah) [inline dashboard/lib/pd/program_registration_validation.rb]: You can require with pegasus_dir instead of this

File: dashboard/app/models/pd/teacher_application.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/models/pd/teacher_application.rb]: I'll use regex:

```ruby
if move_to_user =~ /^\d+$/
```

File: apps/style/code-studio/pd.scss (hunk 3 above)
Comment (aoby) [reply apps/style/code-studio/pd.scss:37]: I'm using the bootstrap 3.0 styles, so yes. I'll move the other styles into the container.

File: dashboard/app/controllers/pd/teacher_application_controller.rb (hunk 1 above)
Comment (Hamms) [reply dashboard/app/controllers/pd/teacher_application_controller.rb:40]: Oh, in that case this is probably fine, particularly for an admin-restricted route.

File: bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb (hunk 7 above)
Comment (aoby) [reply bin/mail/2017-teacher-application-responses/teacher_application_decision_processor.rb:169]: I prefer it this way since it's not returning anything, and to keep the line from being too long.

**Comment by aoby** [comment]
Updated with feedback. PTAL

File: dashboard/app/models/pd/accepted_program.rb [hunk 12]
```diff
@@ -24,6 +25,6 @@ class Pd::AcceptedProgram < ActiveRecord::Base
   TEACHER_CONS_CITY_REGEX = /Houston|Phoenix|Philadelphia/
```
Comment (Hamms) [inline dashboard/app/models/pd/accepted_program.rb]: this regex can be removed, yeah?

File: apps/style/code-studio/pd.scss [hunk 13]
```diff
@@ -42,6 +42,22 @@
       margin: 0;
     }
   }
+
+  table.table.table-auto-width {
```
Comment (mehalshah) [inline apps/style/code-studio/pd.scss]: I think you can still scope this further. Right now this styling would apply to everything in studio because of #application-container

Comment (aoby) [reply apps/style/code-studio/pd.scss]: As it stands, it only applies to tables in studio that have the classes `table` and `table-auto-width`. I made up that last one for this purpose. It seems pretty well scoped to me. Nothing else uses this new class, `table-auto-width`, and the name is descriptive. Also this way it can be used elsewhere if we want it, by applying that class.

Do you have a suggestion to scope it further?
